### PR TITLE
Refactor VMCI Sources

### PR DESCRIPTION
**Pull request by akutz**
This PR replaces PR #1192.

This patch updates the VMCI sources (both C and Go) so that they may be utilized by external actors. The VMCI sources were previously found in these locations:

Directory | File
----------|----
`esx_service/vmci` | `connection_types.h`
 | | `vmci_client.c`
 | | `vmci_server.c`
 | | `vmci_sockets.h`
`vmdk_plugin/drivers/vmdk/vmdkops` | `esx_vmdkcmd.go`

The Go file `esx_vmdkcmd.go` included `vmci_client.c` via Cgo and in doing so linked to the C code statically. However, due to the way Cgo works this meant that external actors that imported the package `github.com/vmware/docker-volume-vsphere/vmdk_plugin/drivers/vmdk/vmdkops` would receive Cgo errors about type redeclaration. This is because the `esx_vmdkcmd.go` file linked directly to a source file instead of a header that used `#pragma once` or `ifdef` statements to guard itself against being included twice.

PR #1192 evolved into a discussion about possible solutions to this issue:

* Create a dynamic, shared library from the C sources which this project and others could utilize by including a `vmci_client.h` header and linking to the shared library. This was rejected due to the need to ensure that the external, shared library would be present on systems with Go programs that required it. It's an external dependency and thus a hassle and unnecessary burden.

* A second option is for external actors to simply copy the C sources to their own projects. This would work unless that project imported the `github.com/vmware/docker-volume-vsphere/vmdk_plugin/drivers/vmdk/vmdkops` package, causing type redeclarations. Additionally this solution would mean that updates to the original sources would not be applied to copies.

@msterin and I agreed that the best and simplest solution is to place the Go source file `esx_vmdkcmd.go` alongside the C sources causing Go to link to it statically but also make it possible to include the C code via a header guarded against redeclaration. 

This patch relocates the following files:

Old | New
----------|----
`esx_service/vmci/connection_types.h` | `vmci/connection_types.h`
`esx_service/vmci/vmci_client.c` | `vmci/vmci_client.c`
 | | `vmci/vmci_client.h`
`esx_service/vmci/vmci_server.c` | `vmci/vmci_server.c`
 | | `vmci/vmci_server.h`
`esx_service/vmci/vmci_sockets.h` | `vmci/vmci_sockets.h`
`vmdk_plugin/drivers/vmdk/vmdkops/esx_vmdkcmd.go` | `vmci/vmci.go`

The Go file `vmdk_plugin/drivers/vmdk/vmdk_driver.go` now imports the package `github.com/vmware/docker-volume-vsphere/vmci` in order to load the ESXi type that satisfies the interface `github.com/vmware/docker-volume-vsphere/vmdk_plugin/drivers/vmdk/vmdkops.VmdkCmdRunner`.

This patch also sanitizes some Go source code in order for it to match de facto as well as de jure standards. Examples include:

* Using `ok` as the boolean flag indicating whether or not a key is a member of a map.
* Using `WithError` instead of including the error as a member of a log entry's fields
* Respecting a maxchar of 80 and wrapping code for readability
* Using natural language for code comments and beginning a comment with the name of the function or field

cc / @msterin, @pdhamdhere

### CODE REVIEW

**Comment by govint** [comment]
@akutz can you also create an issue describing the proposal.

**Comment by msterin** [comment]
@govint - I think this PR has sufficient and well structured information about the reasons, choices and approaches. What would an issue add to that ?

@akutz - thanks for putting it together.  I'll review later today.

File: vmdk_plugin/drivers/vmdk/vmdk_driver.go [hunk 1]
```diff
@@ -194,85 +226,92 @@ func (d *VolumeDriver) UnmountVolume(name string) error {
...
+		return volume.Response{Err: buf.String()}
 	}
 
-	errCreate := d.ops.Create(r.Name, r.Options)
-	if errCreate != nil {
-		log.WithFields(log.Fields{"name": r.Name, "error": errCreate}).Error("Create volume failed ")
-		return volume.Response{Err: errCreate.Error()}
+	if err := d.ops.Create(r.Name, r.Options); err != nil {
+		log.WithField("name", r.Name).WithError(err).Error(
+			"Create volume failed")
```
Comment (govint) [inline vmdk_plugin/drivers/vmdk/vmdk_driver.go]: Nit - can remove white spaces.

File: vmdk_plugin/drivers/vmdk/vmdk_driver.go [hunk 2]
```diff
@@ -194,85 +226,92 @@ func (d *VolumeDriver) UnmountVolume(name string) error {
...
 	// If cloning a existent volume, create and return
-	if _, result := r.Options["clone-from"]; result == true {
-		errClone := d.ops.Create(r.Name, r.Options)
-		if errClone != nil {
-			log.WithFields(log.Fields{"name": r.Name, "error": errClone}).Error("Clone volume failed ")
-			return volume.Response{Err: errClone.Error()}
+	if _, ok := r.Options["clone-from"]; ok {
+		if err := d.ops.Create(r.Name, r.Options); err != nil {
+			log.WithField("name", r.Name).WithError(err).Error(
+				"Clone volume failed")
```
Comment (govint) [inline vmdk_plugin/drivers/vmdk/vmdk_driver.go]: Remove white spaces.

File: vmdk_plugin/drivers/vmdk/vmdk_driver.go (hunk 1 above)
Comment (akutz) [inline vmdk_plugin/drivers/vmdk/vmdk_driver.go]: Hi @govint,

Which white spaces? Any white space you see there is due to [`gofmt`](https://golang.org/cmd/gofmt/) and should remain there if your team and this project adheres to it. I have not yet encountered a Go project that does not since `gofmt` is the official syntax tool for Go.

File: vmdk_plugin/drivers/vmdk/vmdk_driver.go (hunk 2 above)
Comment (akutz) [inline vmdk_plugin/drivers/vmdk/vmdk_driver.go]: Hi @govint,

Which white spaces? Any white space you see there is due to [`gofmt`](https://golang.org/cmd/gofmt/) and should remain there if your team and this project adheres to it. I have not yet encountered a Go project that does not since `gofmt` is the official syntax tool for Go.

**Comment by akutz** [comment]
Hi @msterin (cc @pdhamdhere),

Well color me a fool. I was testing this on the ESX host I stood up for the work I'm doing, and it resulted in the same issue as before. That's when it hit me.

Well, when the other thing hit me. My wife also tossed a pillow at my head around the same time.

I'm an idiot. The reason for the original type redeclaration errors was due to *part* of this project being in my `$GOPATH` and a sub-package from this project being vendored in my project. I am not sure how that happened other than to think Glide didn't pull down all of the sub-packages originally, and since I had already cloned this project into my `$GOPATH`, my build pulled from both locations due to the missing sub-package.

Here is the proof:

```
[root@yellow libstorage]# VMDK_PORT=1019 DRIVERS=vmdk ./got ./drivers/storage/vmw/vmdk/tests -run TestVolumeInspect -v
DEBU[0000] mkdir -p                                      path=/tmp/962894961 time=1493224067383
DEBU[0000] mkdir -p                                      path=/tmp/962894961/etc time=1493224067384
DEBU[0000] mkdir -p                                      path=/tmp/962894961/etc/tls time=1493224067384
DEBU[0000] mkdir -p                                      path=/tmp/962894961/var/lib time=1493224067384
DEBU[0000] mkdir -p                                      path=/tmp/962894961/var/log time=1493224067384
DEBU[0000] mkdir -p                                      path=/tmp/962894961/var/run time=1493224067384
=== RUN   TestVolumeInspect
--- PASS: TestVolumeInspect (0.02s)
	vmdk_test.go:56: status=detached
	vmdk_test.go:56: created by VM=centos7
	vmdk_test.go:56: created=Tue Apr 25 02:52:40 2017
	vmdk_test.go:56: attach-as=independent_persistent
	vmdk_test.go:56: diskformat=thin
	vmdk_test.go:56: clone-from=None
	vmdk_test.go:56: access=read-write
	vmdk_test.go:56: capacity=map[allocated:0 size:100MB]
	vmdk_test.go:56: datastore=datastore1
PASS
ok  	github.com/codedellemc/libstorage/drivers/storage/vmw/vmdk/tests	0.121s

**Comment by akutz** [comment]
Hi @msterin (cc @pdhamdhere),

Well color me a fool. I was testing this on the ESX host I stood up for the work I'm doing, and it resulted in the same issue as before. That's when it hit me.

Well, when the other thing hit me. My wife also tossed a pillow at my head around the same time.

I'm an idiot. The reason for the original type redeclaration errors was due to *part* of this project being in my `$GOPATH` and a sub-package from this project being vendored in my project. I am not sure how that happened other than to think Glide didn't pull down all of the sub-packages originally, and since I had already cloned this project into my `$GOPATH`, my build pulled from both locations due to the missing sub-package.

Here is the proof:

## My Vendored Source
```bash
[root@yellow libstorage]# git --git-dir=vendor/github.com/vmware/docker-volume-vsphere/.git reflog -n1
deeec23 HEAD@{0}: clone: from https://github.com/vmware/docker-volume-vsphere
```

## Not in My `$GOPATH`
```bash
[root@yellow libstorage]# ls $GOPATH/src/github.com 
alecthomas  codedellemc
```

## My Successful Test
```bash
[root@yellow libstorage]# VMDK_PORT=1019 DRIVERS=vmdk ./got ./drivers/storage/vmw/vmdk/tests -run TestVolumeInspect -v
DEBU[0000] mkdir -p                                      path=/tmp/962894961 time=1493224067383
DEBU[0000] mkdir -p                                      path=/tmp/962894961/etc time=1493224067384
DEBU[0000] mkdir -p                                      path=/tmp/962894961/etc/tls time=1493224067384
DEBU[0000] mkdir -p                                      path=/tmp/962894961/var/lib time=1493224067384
DEBU[0000] mkdir -p                                      path=/tmp/962894961/var/log time=1493224067384
DEBU[0000] mkdir -p                                      path=/tmp/962894961/var/run time=1493224067384
=== RUN   TestVolumeInspect
--- PASS: TestVolumeInspect (0.02s)
	vmdk_test.go:56: status=detached
	vmdk_test.go:56: created by VM=centos7
	vmdk_test.go:56: created=Tue Apr 25 02:52:40 2017
	vmdk_test.go:56: attach-as=independent_persistent
	vmdk_test.go:56: diskformat=thin
	vmdk_test.go:56: clone-from=None
	vmdk_test.go:56: access=read-write
	vmdk_test.go:56: capacity=map[allocated:0 size:100MB]
	vmdk_test.go:56: datastore=datastore1
PASS
ok  	github.com/codedellemc/libstorage/drivers/storage/vmw/vmdk/tests	0.121s
```

In other words, I'm a fool who has wasted my time and yours. For that I apologize. I *am*, however, glad that it's working nonetheless.

File: vmdk_plugin/drivers/vmdk/vmdk_driver.go [hunk 3]
```diff
@@ -284,49 +323,53 @@ func (d *VolumeDriver) Create(r volume.Request) volume.Response {
...
 		}
-		errRemove := d.ops.Remove(r.Name, nil)
-		if errRemove != nil {
-			log.WithFields(log.Fields{"name": r.Name, "error": errRemove}).Warning("Remove volume failed ")
+		if err := d.ops.Remove(r.Name, nil); err != nil {
+			log.WithField("name", r.Name).WithError(err).Warn(
+				"Remove volume failed")
 		}
-		return volume.Response{Err: errMkfs.Error()}
+		return volume.Response{Err: err.Error()}
```
Comment (govint) [inline vmdk_plugin/drivers/vmdk/vmdk_driver.go]: Pls. retain the individual errors - errDetach, errRemove. The idea is to return the core failure - mkfs error to the user. The errors that happen when handling the mkfs error to remove the volume are good for logging but isn't the core issue that needs to be returned.

Comment (akutz) [inline vmdk_plugin/drivers/vmdk/vmdk_driver.go]: Hi @govint,

I've closed this PR, but I wanted to let you know that the original errors *are* maintained. The line:

`if err := d.ops.Remove(r.Name, nil); err != nil {`

Does not override the `err` object above it:

`if err := fs.Mkfs(mkfscmd, r.Name, device); err != nil {`

This is due to Go scoping. Here's a quick example in the Go playground ([run](https://play.golang.org/p/f02pmxhXP9)):

```go
package main

import (
	"fmt"
)

func main() {
	x := 0
	if x := 2 + 3; x > 0 {
		fmt.Printf("x=%d\n", x)
	}
	fmt.Printf("x=%d\n", x)
}
```

The above code will print the following:

```
x=5
x=0
```

This is because the `x` in the `if` block is not the same `x` in the main program. It's a new scope.

**Comment by akutz** [comment]
Hi @msterin / @pdhamdhere,

FYI, this is my [test](https://github.com/akutz/libstorage/tree/feature/vmw/drivers/storage/vmdk/tests) package that is using the `HEAD` of this project.

You can run the tests like this from the root of the project (after executing `glide up`):

```bash
$ go test -tags 'pflag gofig libstorage_integration_driver_linux libstorage_storage_driver libstorage_storage_executor libstorage_storage_driver_vmdk libstorage_storage_executor_vmdk' ./drivers/storage/vmdk/tests -run TestVolumeInspect -v
DEBU[0000] mkdir -p                                      path=/tmp/374775760 time=1493226363462
DEBU[0000] mkdir -p                                      path=/tmp/374775760/etc time=1493226363463
DEBU[0000] mkdir -p                                      path=/tmp/374775760/etc/tls time=1493226363463
DEBU[0000] mkdir -p                                      path=/tmp/374775760/var/lib time=1493226363463
DEBU[0000] mkdir -p                                      path=/tmp/374775760/var/log time=1493226363463
DEBU[0000] mkdir -p                                      path=/tmp/374775760/var/run time=1493226363463
=== RUN   TestVolumeInspect
--- PASS: TestVolumeInspect (0.02s)
	vmdk_test.go:56: diskformat=thin
	vmdk_test.go:56: created=Tue Apr 25 02:52:40 2017
	vmdk_test.go:56: clone-from=None
	vmdk_test.go:56: attach-as=independent_persistent
	vmdk_test.go:56: capacity=map[allocated:0 size:100MB]
	vmdk_test.go:56: created by VM=centos7
	vmdk_test.go:56: access=read-write
	vmdk_test.go:56: datastore=datastore1
	vmdk_test.go:56: status=detached
PASS
ok  	github.com/codedellemc/libstorage/drivers/storage/vmdk/tests	0.121s
```

Thank you again for your patience. If it wasn't for the outstanding work on this project, *my* job would be a lot harder. I'm just driving in the final railroad spike after you all have laid the rest of the track. My job is easy, so thank you for making it so!

File: vmdk_plugin/drivers/vmdk/vmdk_driver.go [hunk 4]
```diff
@@ -284,49 +323,53 @@ func (d *VolumeDriver) Create(r volume.Request) volume.Response {
...
-		"fstype": r.Options["fstype"]}).Info("Volume and filesystem created ")
+	log.WithFields(log.Fields{
+		"name":   r.Name,
+		"fstype": r.Options["fstype"],
+	}).Info("Volume and filesystem created")
 	return volume.Response{Err: ""}
 }
 
-// Remove - removes individual volume. Docker would call it only if is not using it anymore
+// Remove - removes individual volume. Docker would call it only if is not
```
Comment (govint) [inline vmdk_plugin/drivers/vmdk/vmdk_driver.go]: Nit - if the volume is not ...."

Comment (akutz) [inline vmdk_plugin/drivers/vmdk/vmdk_driver.go]: Hi @govint,

Please note this PR is closed. Also, you're commenting on something that predates this PR. All I did was wrap the line. The content did not change. If you feel like this should be fixed then I encourage you to fix it.

File: vmdk_plugin/drivers/vmdk/vmdk_driver.go [hunk 5]
```diff
@@ -416,30 +465,29 @@ func (d *VolumeDriver) Mount(r volume.MountRequest) volume.Response {
...
-	log.Debugf("volume name=%s refcnt=%d", r.Name, refcnt)
+
 	if refcnt >= 1 {
-		log.WithFields(
-			log.Fields{"name": r.Name, "refcount": refcnt},
-		).Info("Still in use, skipping unmount request. ")
+		log.WithFields(log.Fields{
+			"name":     r.Name,
+			"refcount": refcnt,
+		}).Debug("volume still in used; skipping unmount request")
```
Comment (govint) [inline vmdk_plugin/drivers/vmdk/vmdk_driver.go]: Lets retain this as Info, the default is the INFO log level and user needs to know if ever a volume has a stuck refcount and refuses to unmount.

Comment (akutz) [inline vmdk_plugin/drivers/vmdk/vmdk_driver.go]: Hi @govint,

Oops, mea culpa. 

Please note I am going to stop responding to your comments. You should stop reviewing this PR. I closed it. So please, stop :)
